### PR TITLE
Add search block

### LINKED ISSUE #1800: Add Block: Search
## Attributes

- Display search icon

## States

Selected:

search

Neutral:

search block

-----

Question:
- I added "Display search icon," but should that be a theme decision, not a user decision?

### PR DESCRIPTION
**Pull request by noisysocks**
Continuation of https://github.com/WordPress/gutenberg/pull/4501.
Closes https://github.com/WordPress/gutenberg/issues/1800.
Implements the design in https://github.com/WordPress/gutenberg/issues/1800#issuecomment-357993628.

Adds a Search block allowing users to place a search field anywhere they like.

search

- Users can customise the label by editing it inline
- Users can customise the placeholder text using the block inspector 
- Our own custom markup is used instead of `get_search_form()`, see https://github.com/WordPress/gutenberg/pull/4501#pullrequestreview-97716621 for context

Props to @bamadesigner for all her hard work on this in https://github.com/WordPress/gutenberg/pull/4501.

#### To test

1. Create a new post
1. Insert the Search block
1. Publish the post
1. Type a query into the search field you just added
1. Press <kbd>Enter</kbd>

### CODE REVIEW

**Review by Soean — DISMISSED** [review]
Just did a short test, works great. Just one improvement: If I add a custom class, it is not added to the form.

File: packages/block-library/src/search/edit.js
Comment (gziolo) [inline packages/block-library/src/search/edit.js]: Should it be also using `form` to ensure that CSS is properly applied if someone uses a tag name specific selectors?

**Review by gziolo — DISMISSED** [review]
Let's also update `CHANGELOG.md` and request minor version update for the next package release.

We discussed that yesterday during Core JS weekly chat: https://make.wordpress.org/core/2019/01/29/javascript-chat-summary-january-29-2019/

@Soean - I realized we didn't do it for RSS block, so it would be good to open a separate PR and fix that.

File: packages/block-library/src/search/edit.js
Comment (gziolo) [inline packages/block-library/src/search/edit.js]: @melchoyce or @mapk - how would you feel about using inline editing for the placeholder? We would have to style `RichText` to look like an input field but it might be closer to the principles that blocks approach promotes.

**Comment by gziolo** [comment]
Yes, it is close to being ready in my opinion. If we apply changes by the end of the week we can move it to 5.0 (Gutenberg) milestone 🎉

**Review by talldan — COMMENTED** [review]
I had a couple of minor nitpicks, but this looks good 😄

File: packages/block-library/src/search/edit.js
Comment (talldan) [inline packages/block-library/src/search/edit.js]: I think should have an aria-label—other blocks seem to achieve that by using the `placeholder` prop. Also perhaps `multiline={ false }`?

**Comment by gziolo** [comment]
@michelleweber would you mind helping to polish translations proposed?

File: packages/block-library/src/search/edit.js
Comment (gziolo) [reply packages/block-library/src/search/edit.js]: See also my proposal which is an alternative approach to consider: https://github.com/WordPress/gutenberg/pull/13583#discussion_r252183078

**Comment by afercia** [comment]
Adding the Accessibility label in addition to the "needs..." one so the team can search for all the widget-blocks related issues in an easier way.

File: packages/block-library/src/search/edit.js
Comment (gziolo) [reply packages/block-library/src/search/edit.js]: Also sharing this comment from WordPress.org Slack from @afercia https://wordpress.slack.com/archives/C02RP4X03/p1548839191742400 (link requires registration) and in particular this concern raised:
>  the widget-blocks have important settings placed in the sidebar

**Comment by afercia** [comment]
It would be great for this block to not encourage the use of placeholders as replacement for labels, see https://core.trac.wordpress.org/ticket/40331

So instead of a default `Enter a search keyword or phrase` I'd consider to either:
- leave it empty by default and keep the placeholder setting in the sidebar (even if the sidebar is not ideal for the reason @gziolo mentioned)
- use a default `Optional placeholder...` which of course is not rendered in the front end, and explore to make the field editable as @gziolo proposed, thus removing the setting from the sidebar

More importantly:
I'm not sure why this block is removing the Search **button**. The classic Search widget prints out a search button and that's recommended and expected.

On top: the new search block, followed by the classic search widget:

screenshot 2019-01-30 at 14 54 26

For accessibility a search form should always have a button

**Comment by michelleweber** [comment]
Sure, you can specify what bits exactly you'd like reviewed?

File: packages/block-library/src/search/edit.js
Comment (melchoyce) [reply packages/block-library/src/search/edit.js]: I think we should give it a try — going to be hard to judge until we can test it in-context.

**Comment by melchoyce** [comment]
Looking at some major sites (Google, Amazon, Twitter, Amazon) their search fields don't have a visible label — are they doing something else to make it a11y-compliant? 

Quick idea to include a button, based on how some of the aforementioned sites do it:

image

**Comment by ZebulanStanphill** [comment]
@melchoyce They may be putting `aria-label`s on their search bars. However, this is less-than-ideal; visible labels should always be preferred.

On that note, I think the search button should have a visible label as well... actually, why not allow the user to edit the button text as well?

**Comment by melchoyce** [comment]
If the search button had a visible label, what would be ideal? I don't think it makes sense for us to use Search as an input label, and a button label.

**Comment by afercia** [comment]
I'd tend to think the visible label should be in the hands of the authors. Gutenberg shouldn't force a search field with no visible, properly associated, `<label>` element over an `aria-label`. That would mean forcing a decreased level of accessibility. 

If authors opt to not have a visible label, then that's their responsibility.

Ideally, in the same way Gutenberg does for the headings hierarchy and color contrast, a visible label should be the recommended option. Authors should be educated, not forced to follow an opinionated choice.

**Comment by ZebulanStanphill** [comment]
@melchoyce I guess "Submit", "Go", or "Enter" would be fine defaults if "Search" isn't used for the button label. Alternatively, the button could be labeled "Search" and the label preceding the search input could be "Search this site" or something like that by default. Whatever the case, both the preceding label and button should be editable.

**Comment by melchoyce** [comment]
Okay — so it sounds like we'll want an inline editable label, placeholder, and search button.

@michelleweber What do you think the defaults should be? For example:

**Input label**: Search
**Placeholder**: Enter a search keyword or phrase
**Button Label**: Submit

**Comment by melchoyce** [comment]
I understand that placeholders shouldn't replace labels, but why should we discourage them altogether?

**Comment by afercia** [comment]
We should try to discourage their use as replacements for labels 🙂 
Instead, if they add some hint on the expected value or format, [they're perfectly fine](https://www.w3.org/TR/html52/sec-forms.html#the-placeholder-attribute).

**Comment by skorasaurus** [comment]
Just my two cents from building one:

- placeholders often have lighter text which doesn't meet contrast requirements for accessibility;  
- on mobile devices, placeholder text can be truncated; 
For example, on an iphone 6 (simulated screen size w/in firefox); the placeholder text isn't even complete; 

It's not any one thing that's ; but a combination of things that are often viewed as edge cases that together; make them more difficult the user experience more difficult than it's worth of using them. 

as @afercia mentioned cited in a core ticket, [smashing magazine](https://www.smashingmagazine.com/2018/06/placeholder-attribute/) and [nielsen group ](https://www.nngroup.com/articles/form-design-placeholders/) outline why not to use them which I had mentioned

screen shot 2019-01-30 at 13 57 23

**Comment by melchoyce** [comment]
Thanks 👍 

How's this?

image

**Comment by michelleweber** [comment]
I don't think you need the word "search" in the placeholder (if you're going to use the placeholder).

For the label and button, I like "Search this site" and "Search" (I always like a more active button where possible).

File: packages/block-library/src/search/edit.js
Comment (noisysocks) [reply packages/block-library/src/search/edit.js]: Is that likely? We already have the problem of the `<label>` being a `<div>` because of some `RichText`. Making this a `<form>` means that we'd have to add a `onSubmit={ ( event ) => event.stopPropagation() }` which I find a little distracting.

File: packages/block-library/src/search/edit.js
Comment (noisysocks) [reply packages/block-library/src/search/edit.js]: Good catch—added the `placeholder` prop so that there's a label.

https://github.com/WordPress/gutenberg/issues/13570 is a ticket that tracks adding `multiline={ false }` which looks like it was removed a while ago. For now, I think it's fine to support line breaks—I can't think of any reason we ought to forbid them! 🙂

**Comment by noisysocks** [comment]
Thanks for the quick review, friends! In 4f4da5e I have:

- Fixed custom CSS classes not working
- Made the placeholder text editable inline
- Made there not be a placeholder by default
- Added a search button with text that is editable inline

Here's how it's looking:

search

File: packages/block-library/src/search/edit.js
Comment (talldan) [reply packages/block-library/src/search/edit.js]: Was sure I could remember `multiline={ false }` working in the editor, but I must have dreamed it 😴. My feeling is that it's  bit unusual for a label to be multiple lines.

Thanks for the fixes, will re-review in a bit.

File: packages/block-library/src/search/edit.js
Comment (gziolo) [reply packages/block-library/src/search/edit.js]: I don't know what is the best way moving forward, I just wanted to raise this so we could discuss and come up with something solid as this is going to be replicated in other places :)

File: packages/block-library/CHANGELOG.md
Comment (gziolo) [inline packages/block-library/CHANGELOG.md]: Cool, thanks 👍

**Comment by gziolo** [comment]
@noisysocks, the screencast you shared looks great in my opinion. It should be much easier to customize it now using keyboard only. I'll test it later today,

**Comment by gziolo** [comment]
I'm also sharing for reference Search form from my website which uses Twenty Fifteen theme:
screen shot 2019-01-31 at 10 16 29

It uses [class screen-reader-text](https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/) to keep it accessible but hides both label and submit button as a default experience. We should consider a follow-up PR where we enable those options in the sidebar.

File: packages/block-library/src/search/edit.js
Comment (talldan) [inline packages/block-library/src/search/edit.js]: Should this also be a RichText? I suppose it's ok as an input.

Either way, I think it should have an `aria-label`, as placeholders aren't considered a replacement for labels.

File: packages/block-library/src/search/style.scss
Comment (talldan) [inline packages/block-library/src/search/style.scss]: `font-weight: bold` might be a bit of an opinionated style that themers will need to override. 

Moving it to the editor.scss and theme.scss might be a good idea.

**Comment by afercia** [comment]
Thanks for working on this 🙂 

Noticed a few things:

1
The label field and the button are RichText components and thus are wrapped in a `<Autocomplete>` component. For example, type `@`  in both of them and you get the user mentions suggestions.

screenshot 2019-01-31 at 09 27 46

Also, by default they have an `aria-autocomplete="list"` attribute. I realize this depends on `RichText`. To my understanding, passing `autocompleters={ [] }` disables the suggestions but doesn't remove the `aria-autocomplete="list"` attribute and the wrapper.  Maybe `RichText` should be changed with a prop to not render the auto-completer stuff. Should this go in a separate issue?

2
This is a bit tricky: when saving a custom placeholder text, it gets saved as the input `value`. It would be great to save it as the `placeholder` but still make it behave like an editable `value`. Otherwise the saved and rendered markup will be something like this:
`placeholder="Optional placeholder" value="My custom placeholder"`

which semantically doesn't represent correctly what the UI communicates. Screen readers will announce the value as "content" of the field, together with the default placeholder:

screenshot 2019-01-31 at 10 45 53

3
The search field is technically unlabelled. It needs an `aria-label` but I'm not sure what an appropriate aria-label value would be in this editing context. Maybe `aria-label="Optional placeholder text"` ?

4
In the label field and in the button: select all the text and delete it. The value used for the aria-label is then displayed as the default placeholder:

screenshot 2019-01-31 at 09 14 54

5
MInor but worth mentioning: usually, default placeholders have a horizontal ellipsis at the end, e..g. `__( 'Write caption…' )` so I'd suggest to use `__( 'Optional placeholder…' )`

**Comment by gziolo** [comment]
> This is a bit tricky: when saving a custom placeholder text, it gets saved as the input `value`. It would be great to save it as the `placeholder` but still make it behave like an editable `value`. Otherwise the saved and rendered markup will be something like this:
> `placeholder="Optional placeholder" value="My custom placeholder"`
> 
> which semantically doesn't represent correctly what the UI communicates. Screen readers will announce the value as "content" of the field, together with the default placeholder:

This block is server-side rendered and `placeholder` gets saved as `placeholder`. It only works this way in the editor. What issues do you see with this approach?

**Comment by afercia** [comment]
@gziolo in the editor it gets saved as `placeholder="Optional placeholder" value="My custom placeholder"` so when editing, assistive technologies users will be very confused.

File: packages/block-library/src/search/style.scss
Comment (ZebulanStanphill) [reply packages/block-library/src/search/style.scss]: Agreed. We should be avoiding opinionated styles as much as possible for core blocks.

File: packages/block-library/src/search/edit.js
Comment (noisysocks) [reply packages/block-library/src/search/edit.js]: Line breaks and HTML aren't supported in the `placeholder` attribute so I went with `<input>` here instead of `PlainText` or `RichText`.

File: packages/block-library/src/search/edit.js
Comment (noisysocks) [reply packages/block-library/src/search/edit.js]: You didn't dream it! 😀 I think that we removed support for `multiline={ false }` when refactoring how `RichText` works in the lead up to 5.0.

I did a little audit of line breaks in the Search block's attributes and I'm OK with how it is currently:

- `label`: `<br>`s are allowed in a `<label>` so we permit it. This is consistent with e.g. captions in the Image block.
- `placeholder`: `<br>`s or `\n`s are not permitted in a `placeholder=""` so we forbid it.
- `buttonText`: `<br>`s are allowed in a `<button>` so we permit it. This is consistent with e.g. the button text in a Button or File block.

File: packages/block-library/src/search/edit.js
Comment (noisysocks) [reply packages/block-library/src/search/edit.js]: Added an `aria-label` in 1eae8ea6c! 👌

File: packages/block-library/src/search/style.scss
Comment (noisysocks) [reply packages/block-library/src/search/style.scss]: Moved it to `style.scss` in fb9b79015! 👌

**Comment by noisysocks** [comment]
> The label field and the button are RichText components and thus are wrapped in a <Autocomplete> component. For example, type @ in both of them and you get the user mentions suggestions.

I'll admit it's a little bizarre but this is consistent with how other blocks work.

For example, one can mention someone in the caption of an Image block:

image mention

Similarly, one can mention someone in the text of a Button block:

button mention

If this behaviour isn't desired then I suggest we ticket it separately.

> This is a bit tricky: when saving a custom placeholder text, it gets saved as the input value. It would be great to save it as the placeholder but still make it behave like an editable value.

I don't _think_ that this is technically possible but I'll experiment with some options and report back.

> The search field is technically unlabelled. It needs an aria-label but I'm not sure what an appropriate aria-label value would be in this editing context. Maybe aria-label="Optional placeholder text" ?

👌 Good catch, I've added a label in 1eae8ea6cf3fa1b11a9cd01e097b9b8aa40e80df.

> In the label field and in the button: select all the text and delete it. The value used for the aria-label is then displayed as the default placeholder:

👌 I've added explicit `aria-label`s for all of the fields in 0d3fb096f.

> MInor but worth mentioning: usually, default placeholders have a horizontal ellipsis at the end, e..g. __( 'Write caption…' ) so I'd suggest to use __( 'Optional placeholder…' )

👌 Good catch, was confused about this since _Add title_ at the top of a post doesn't have an ellipsis. Changed in d87c67dee.

**Comment by noisysocks** [comment]
> I don't think that this is technically possible but I'll experiment with some options and report back.

I've made the placeholder text input use `placeholder` instead of `value` in 154cddcac. If I'm honest, I don't understand at all _why_ this works but it does (even in IE10!) so chalking this up as a win! 🤷‍♂️

**Comment by noisysocks** [comment]
I'm thinking this is _really_ close so moving it on up to the 5.0 milestone 🚗💨

**Comment by noisysocks** [comment]
> This is a bit tricky: when saving a custom placeholder text, it gets saved as the input value. It would be great to save it as the placeholder but still make it behave like an editable value.

After thinking about this some more and talking it over with @talldan, I'm convinced that this isn't what we want. 

I've reverted 154cddc so that in the editor, the UI for editing these three attributes (`label`, `placeholder` and `buttonText`) behaves as though there are three inputs with the following labels and placeholders:

- Label text: <Insert label text…>
- Placeholder text: <Optional placeholder…>
- Button text: <Insert button text…>

Gutenberg being very WYSIWYG-ey means that it's easy to get confused and think that the markup in the editor must match the markup in the post, when in fact the editor markup should only be _styled_ to look like what is in the post.

**Comment by afercia** [comment]
Yep I realize the placeholder saving is tricky. However, the way it works now is semantically incorrect and misleading for users who perceive the placeholder content. Consider this example:

screenshot 2019-02-01 at 09 13 41

The actual markup _in the editor_ would contain these values:
```
aria-label="Optional placeholder text"
placeholder="Optional placeholder…"
value="Find Movies, TV shows, Celebrities and more..."
```

All these three values are announced by screen readers. The default placeholder text will still be announced as such because... well, it is the placeholder. Please see below: it's a bit confusing 🙂

screenshot 2019-02-01 at 09 21 14

I understand in an editing context things are a bit different and it's not always possible to be 100% semantic.

If it is not possible / desirable to save "Find Movies..." as the actual placeholder value, would it be possible to remove the placeholder attribute when there is a value? Not saying it would be correct semantically but at least it would be a bit less confusing.

**Review by talldan — APPROVED** [review]
I like it.

Just one more thing (Columbo style) - An observation is that I found it a bit unusual that when adding the block the caret was placed at the start of the 'Search' text instead of the end. 

I imagine what happens here is the same (writing flow) behaviour as when using the down arrow key to navigate into a paragraph block that already has some text. That behaviour makes sense since the user is essentially navigating between lines of text.

I also couldn't find other examples of blocks that have default values like this, so it might be a new thing. Not a showstopper, IMO.

**Comment by afercia** [comment]
Can the point I've raised above be addressed before approving please? Thank you 🙂

**Comment by talldan** [comment]
> Can the point I've raised above be addressed before approving please? Thank you

@afercia Please can you use the github feature for 'Requesting changes'. Merging will be blocked while there are requested changes as it is now due to the other reviews.

> However, the way it works now is semantically incorrect and misleading for users who perceive the placeholder content.

This is a really confusing one, because it's I think it's the first example we have of an input that is present when viewing the post that also has a counterpart in the editor. 

The important bit though is that the input when viewing the post is semantically and contextually different to the input present when editing the post:
- The input when viewing the post is used to enter search criteria
- The input when editing the post is used for editing a placeholder value (but the key bit is that it's not editing its own placeholder value, but instead the placeholder value on the input when you view the post).

If as in your example, the placeholder on the input in the editor were updated to 'Find Movies, TV shows, Celebrities and more...', this would lead the user to think that typing a value in that editor's input field would enable them to search for movies, but that's not the case.

**Comment by afercia** [comment]
> Please can you use the github feature for 'Requesting changes'.

Will try. I tend to not use it because I feel like it's a bit intrusive.

> The important bit though is that the input when viewing the post is semantically and contextually different to the input present when editing the post

Absolutely. That's the point when I've referred to "editing context" (or "authoring context"). There are precedents, for example the images alt attribute, see #11218.

Basically: what is the correct information users need in an editing context? In some cases, it's different from the one they need in the front end.

**Review by afercia — DISMISSED** [review]
Please see previous comment about the placeholder value.

**Comment by gziolo** [comment]
> > Please can you use the github feature for 'Requesting changes'.
> 
> Will try. I tend to not use it because I feel like it's a bit intrusive.

It shouldn't feel like that. It gives a clear message that there are issues that are blocking merging PR.

**Comment by talldan** [comment]
> There are precedents, for example the images alt attribute

I don't believe it's the same thing. The image alt value usage is correct, because if I upload an image of say, a house, and set the alt as "Image of a house", the alt value is correct in describing the image.

In this case, if the placeholder in the editor says "Search for films", it's incorrect because the user cannot search for films.

**Comment by afercia** [comment]
Right. The user can't search at all. But also the _value_ "Search for films" is incorrect. Because the user can't search for the terms "Search for films" 🙂 

How this could be explained in plain words? Ideally, it should make clear "this is the placeholder text that will be used in the front end". But it's getting a bit complicated 🙂

**Comment by noisysocks** [comment]
> If it is not possible / desirable to save "Find Movies..." as the actual placeholder value, would it be possible to remove the placeholder attribute when there is a value? Not saying it would be correct semantically but at least it would be a bit less confusing.

In f3475f5ff I've changed the placeholder input so that it's `placeholder` property is removed when there is a `value`. The input still has an `aria-label` of `'Optional placeholder text'` which I think is pretty clear.

**Comment by mapk** [comment]
✅ Just tested, and it feels and works great from a design view. I'm able to delete the label in my search block which is cool. The placeholder worked as expected. The button editing was smooth as can be. Thanks for pushing this forward!

**Comment by gziolo** [comment]
@Soean and @afercia - it looks like your issues were resolved, can you review again?

**Review by gziolo — APPROVED** [review]
Let's give some time @afercia and @Soean to confirm their feedback was addressed and let's get this in 🎉

**Comment by afercia** [comment]
Looks better from an accessibility perspective, thank you. Note: I see a warning:
`Warning: Received `false` for a non-boolean attribute `placeholder`.`

because the placeholder gets actually evaluated to `false`. Should probably use a ternary that returns `undefined` when the input has value.

**Comment by noisysocks** [comment]
Good catch @afercia! Fixed in dffa8d380.

**Comment by noisysocks** [comment]
Will merge this as it has two approvals and the remaining issues identified by @afercia and @Soean have been addressed. As always, we can use follow-up PRs to make further changes.

Thanks all for reviewing! And thanks to @bamadesigner for getting this ball rolling! ✌️❤️

File: test/integration/full-content/fixtures/core__search__custom-text.serialized.html
Comment (aduth) [inline test/integration/full-content/fixtures/core__search__custom-text.serialized.html]: The fact that this markup does not match `core__search__custom-text.html` should be alarming; it's the entire purpose of these fixtures to guarantee. Running `npm run fixtures:regenerate` locally seems to produce some changes for me which produce a result more to what's expected.

However, it's still very concerning that:

- The build passes despite these markups being entirely different
- The build passes despite there being local changes to commit after running `npm run fixtures:regenerate`

Comment (aduth) [reply test/integration/full-content/fixtures/core__search__custom-text.serialized.html]: See #14122 for fix

**Comment by JoshuaJarman** [comment]
sorry to have to ask here, but searched everywhere.  which block does one use to display the search results?  block template link to the search slug renders but which block do we use to display the paginated search results?  thanks.

**Comment by noisysocks** [comment]
There isn't a block to display search results, yet. Right now the search block will take the user to a page which uses the search WordPress template.

**Comment by JoshuaJarman** [comment]
i see.  thanks.

we are building a number of pure gutenberg only sites right now, using the block wp_template to theme template mapping so that the only php template file we have is a functions file, no template files other than that, everything else is pure blocks and block templates.

we have every feature working on all the sites so far using only blocks except search results.  looks like we'll have to write a custom block for that one last piece.  pure gutenberg sites are soooo close, this is the last piece.  👍

**Comment by noisysocks** [comment]
> we have every feature working on all the sites so far using only blocks except search results. looks like we'll have to write a custom block for that one last piece. pure gutenberg sites are soooo close, this is the last piece. 👍

Agree they're close!

After you've built a custom search results block, I'd strongly encourage you to contribute it to Core by opening a PR in this repository. WordPress is better together! 💪

**Pull request merged**
